projects/kit/components/input-range/input-range.component.ts

```diff
@@ -0,0 +1,216 @@
+            this.range.legacyMode = false; // TODO(v5): remove backward compatibility
+        }
+    }
+
+    protected get hideStartContent(): boolean {
+        return (
+            !this.content[0] ||
+            tuiIsNativeFocused(this.textfieldStart) ||
+            (tuiIsNativeFocusedIn(this.rangeRef?.nativeElement) &&
+                this.lastActiveSide === 'start')
+        );
+    }
+
+    protected get hideEndContent(): boolean {
+        return (
+            !this.content[1] ||
+            tuiIsNativeFocused(this.textfieldEnd) ||
+            (tuiIsNativeFocusedIn(this.rangeRef?.nativeElement) &&
+                this.lastActiveSide === 'end')
+        );
+    }
+
+    protected takeStep(
+        event: Event | KeyboardEvent,
+        coefficients: readonly [number, number],
+    ): void {
+        if (!this.interactive() || !this.range) {
+            return;
+        }
+
+        event.preventDefault();
+
+        const [start, end] = this.value();
+        const newValue = this.valueGuard(this.range.takeStep(coefficients));
+
+        if (newValue[0] !== start || newValue[1] !== end) {
+            this.onExternalValueUpdate(newValue);
+        }
+    }
+
+    protected onInput([start, end]: [number | null, number | null]): void {
+        const [prevStart, prevEnd] = this.value();
+
+        this.setValue([start ?? prevStart, end ?? prevEnd]);
+    }
+
+    protected onExternalValueUpdate(value: readonly [number, number]): void {
+        this.setValue(value);
+        this.setTextfieldValues(this.value());
+    }
+
+    protected focusToTextInput(): void {
+        const element =
+            this.lastActiveSide === 'start' ? this.textfieldStart : this.textfieldEnd;
+
+        if (!this.isMobile && element) {
+            element.focus();
+        }
+    }
+
+    protected onActiveThumbChange(activeThumb: 'left' | 'right'): void {
+        // TODO(v5): remove backward compatibility
+        this.lastActiveSide = activeThumb === 'left' ? 'start' : 'end';
+    }
+
+    protected setTextfieldValues([start, end]: readonly [number, number]): void {
+        this.textfieldValueStart = start;
+        this.textfieldValueEnd = end;
+    }
+
+    private get textfieldStart(): HTMLInputElement | null {
+        return this.inputNumberRefs.first?.nativeElement || null;
+    }
+
+    private get textfieldEnd(): HTMLInputElement | null {
+        return this.inputNumberRefs.last?.nativeElement || null;
+    }
+
+    private setValue(value: readonly [number, number]): void {
+        this.onChange(this.valueGuard(value));
+    }
+
+    private valueGuard([start, end]: readonly [number, number]): readonly [
+        number,
+        number,
+    ] {
+        const [prevStart, prevEnd] = this.value();
+
+        return [
+            Math.min(this.calibrate(start), prevEnd),
+            Math.max(this.calibrate(end), prevStart),
+        ];
+    }
+
+    private calibrate(value: number): number {
+        return tuiClamp(
```

Comment:
Do we really need `tuiClamp` (if `InputNumber` + `Range` validates it)? Check it